transports/quic/src/transport.rs

```diff
@@ -0,0 +1,639 @@
+    /// Dialer for Ipv4 addresses if no matching listener exists.
+    ipv4_dialer: Option<Dialer>,
+    /// Dialer for Ipv6 addresses if no matching listener exists.
+    ipv6_dialer: Option<Dialer>,
+
+    _marker: PhantomData<P>,
+}
+
+impl<P> QuicTransport<P> {
+    pub fn new(config: Config) -> Self {
+        Self {
+            listeners: SelectAll::new(),
+            config,
+            ipv4_dialer: None,
+            ipv6_dialer: None,
+            _marker: Default::default(),
+        }
+    }
+}
+
+/// Error that can happen on the transport.
+#[derive(Debug, thiserror::Error)]
+pub enum TransportError {
+    /// Error while trying to reach a remote.
+    #[error("{0}")]
+    Reach(quinn_proto::ConnectError),
+    /// Error after the remote has been reached.
+    #[error("{0}")]
+    Established(ConnectionError),
+
+    #[error("{0}")]
+    Io(#[from] std::io::Error),
+
+    /// The [`EndpointDriver`](super::endpoint::EndpointDriver) has crashed.
+    #[error("Endpoint driver crashed")]
+    EndpointDriverCrashed,
+}
+
+impl<P: Provider> Transport for QuicTransport<P> {
+    type Output = (PeerId, QuicMuxer);
+    type Error = TransportError;
+    type ListenerUpgrade = Upgrade;
+    type Dial = Pin<Box<dyn Future<Output = Result<Self::Output, Self::Error>> + Send>>;
+
+    fn listen_on(
+        &mut self,
+        addr: Multiaddr,
+    ) -> Result<ListenerId, CoreTransportError<Self::Error>> {
+        let socket_addr = multiaddr_to_socketaddr(&addr)
+            .ok_or(CoreTransportError::MultiaddrNotSupported(addr))?;
+        let listener_id = ListenerId::new();
+        let listener = Listener::new::<P>(listener_id, socket_addr, self.config.clone())
+            .map_err(CoreTransportError::Other)?;
+        self.listeners.push(listener);
+        // Drop reference to dialer endpoint so that the endpoint is dropped once the last
+        // connection that uses it is closed.
+        // New outbound connections will use a bidirectional (listener) endpoint.
+        match socket_addr {
+            SocketAddr::V4(_) => self.ipv4_dialer.take(),
+            SocketAddr::V6(_) => self.ipv6_dialer.take(),
+        };
+        Ok(listener_id)
+    }
+
+    fn remove_listener(&mut self, id: ListenerId) -> bool {
+        if let Some(listener) = self.listeners.iter_mut().find(|l| l.listener_id == id) {
+            listener.close(Ok(()));
+            true
+        } else {
+            false
+        }
+    }
+
+    fn address_translation(&self, _server: &Multiaddr, observed: &Multiaddr) -> Option<Multiaddr> {
+        Some(observed.clone())
+    }
+
+    fn dial(&mut self, addr: Multiaddr) -> Result<Self::Dial, CoreTransportError<Self::Error>> {
+        let socket_addr = multiaddr_to_socketaddr(&addr)
+            .ok_or_else(|| CoreTransportError::MultiaddrNotSupported(addr.clone()))?;
+        if socket_addr.port() == 0 || socket_addr.ip().is_unspecified() {
+            return Err(CoreTransportError::MultiaddrNotSupported(addr));
```

Comment:
It might be time to go through all the transports and just see if we can add more variants that cover more specific cases.

Reply:
It might be worth splitting this `if` into two so we can add a comment that support for wildcard ports will be added eventually.

Reply:
Could you give examples of what specific reasons/ cases you have in mind @mxinden @thomaseizinger? Did we have situations in the past where users had this error and struggled with debugging it?

> It might be worth splitting this if into two so we can add a comment that support for wildcard ports will be added eventually.

@thomaseizinger this code is part of `Transport::dial`, so we will never support dialing wildcard ports here. Am I missing something?

Reply:
> Could you give examples of what specific reasons/ cases you have in mind @mxinden @thomaseizinger? Did we have situations in the past where users had this error and struggled with debugging it?

_Just to make sure I am not creating any confusion, I think this is outside of the scope of this pull request._

I have had the problem multiple times where I pass a malformed `Multiaddr` to a hierarchy of `Transport` implementations (e.g. relayed addresses are tricky) and my `Multiaddr` is being denied via `MutliaddrNotSupported`.

Say I have the following address: `/dns6/xxx/tcp/xxx/p2p/xxx/p2p-circuit/p2p/xxx`. When I get a `MultiaddrNotSupported` I don't know whether this is due to:

- The address being malformed for `libp2p-relay`.
- The address being malformed for `libp2p-dns`.
- The addresses resolved by `libp2p-dns` being malformed for `libp2p-tcp`.

In addition to not knowing where it has been denied, I don't know what the reason for denial was.

Reply:
I created https://github.com/libp2p/rust-libp2p/issues/2980 to track the `MultiaddrNotSupported` suggestion. Leaving this open for the second discussion, namely:

> > It might be worth splitting this if into two so we can add a comment that support for wildcard ports will be added eventually.
> 
> @thomaseizinger this code is part of `Transport::dial`, so we will never support dialing wildcard ports here. Am I missing something?

Reply:
Thanks for elaborating @mxinden. Since this is tracked in #2980 and out of scope for this PR I mark it as resolved.